Readme.md
# retext-double-metaphone [![Build Status](https://img.shields.io/travis/wooorm/retext-double-metaphone.svg?style=flat)](https://travis-ci.org/wooorm/retext-double-metaphone) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-double-metaphone.svg?style=flat)](https://coveralls.io/r/wooorm/retext-double-metaphone?branch=master)

**[Retext](https://github.com/wooorm/retext)** implementation of the [Double Metaphone](http://en.wikipedia.org/wiki/metaphone) algorithm.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-double-metaphone
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-double-metaphone
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-double-metaphone
```

[Duo](http://duojs.org/#getting-started):

```javascript
var doubleMetaphone = require('wooorm/retext-double-metaphone');
```

## Usage

```javascript
var Retext = require('retext');
var visit = require('retext-visit');
var inspect = require('retext-inspect');
var doubleMetaphone = require('retext-double-metaphone');

var retext = new Retext()
    .use(inspect)
    .use(visit)
    .use(doubleMetaphone);

retext.parse('A simple english sentence.', function (err, tree) {
    tree.visit(tree.WORD_NODE, function (node) {
        console.log(node);
    });
    /*
     * WordNode[1] [data={"phonetics":["A","A"]}]
     * └─ TextNode: 'A'
     * WordNode[1] [data={"phonetics":["SMPL","SMPL"]}]
     * └─ TextNode: 'simple'
     * WordNode[1] [data={"phonetics":["ANKLX","ANLX"]}]
     * └─ TextNode: 'english'
     * WordNode[1] [data={"phonetics":["SNTNS","SNTNS"]}]
     * └─ TextNode: 'sentence'
     */
});
```

You can also combine it with a stemmer (such as [retext-porter-stemmer](https://github.com/wooorm/retext-porter-stemmer), [retext-lancaster-stemmer](https://github.com/wooorm/retext-lancaster-stemmer)).

```javascript
var Retext = require('retext');
var visit = require('retext-visit');
var inspect = require('retext-inspect');
var doubleMetaphone = require('retext-double-metaphone');
var stemmer = require('retext-porter-stemmer');

var retext = new Retext()
    .use(inspect)
    .use(visit)
    .use(doubleMetaphone)
    /* make sure to attach the stemmer after metaphone. */
    .use(stemmer);

retext.parse('A detestable paragraph.', function (err, tree) {
    tree.visit(tree.WORD_NODE, function (node) {
        console.log(node);
    });
    /*
     * WordNode[1] [data={"stem":"a","phonetics":["A","A"],"stemmedPhonetics":["A","A"]}]
     * └─ TextNode: 'A'
     * WordNode[1] [data={"stem":"detest","phonetics":["TTSTPL","TTSTPL"],"stemmedPhonetics":["TTST","TTST"]}]
     * └─ TextNode: 'detestable'
     * WordNode[1] [data={"stem":"paragraph","phonetics":["PRKRF","PRKRF"],"stemmedPhonetics":["PRKRF","PRKRF"]}]
     * └─ TextNode: 'paragraph'
     */
});
```

## API

None, **retext-double-metaphone** automatically detects the phonetics of each [`WordNode`](https://github.com/wooorm/textom#textomwordnode-nlcstwordnode) (using **[wooorm/double-metaphone](https://github.com/wooorm/double-metaphone)**), and stores the phonetics in `node.data.phonetics`. If a stemmer is used, the stemmed phonetics are stored in `node.data.stemmedPhonetics`.

## Performance

On a MacBook Air, **retext** performs about 20% slower with **retext-double-metaphone**.

```text
           retext w/o retext-double-metaphone
  217 op/s » A paragraph (5 sentences, 100 words)
   24 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)

           retext w/ retext-double-metaphone
  173 op/s » A paragraph (5 sentences, 100 words)
   16 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
```

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md
<!--mdast setext-->

<!--lint disable no-multiple-toplevel-headings-->

<!--lint disable maximum-line-length-->

0.1.7 / 2015-01-19
==================

*   Update copyright notice in `LICENSE` to include 2015
*   Add link to whole license in `Readme.md`
*   Add Duo as an instalation method in `Readme.md`
*   Add links to installation methods in `Readme.md`
*   Refactor fences code blocks in `Readme.md`
*   Refactor to adhere to strict jsdoc style
*   Add jscs-jsdoc configuration to `.jscs.json`
*   Add jscs-jsdoc as a dev-dependency
*   Refactor npm scripts for changes in npm
*   Update eslint
*   Refactor to use generic `changeinside` in new retext
*   Update retext to stable release

0.1.6 / 2014-12-13
==================

*   Add benchmark
*   Fix wording in `Readme.md`
*   Add link to personal website to copyright in `Readme.md`
*   Update eslint, retext

0.1.5 / 2014-11-21
==================

*   Refactor `Readme.md`
*   Add flat badges to `Readme.md`
*   Fix incorrect executive rights on `test.js`
*   Refactor npm script targets in `package.json`
*   Add retext-inspect to dev-dependecies
*   Update retext

0.1.4 / 2014-10-28
==================

*   Add `.eslintrc`
*   Refactor to disallow space after object keys
*   Refactor for changes in retext
*   Update eslint, mocha, retext, retext-visit, retext-porter-stemmer, retext-content

0.1.3 / 2014-10-21
==================

*   Refactor .jscs.json
*   Update retext

0.1.2 / 2014-10-14
==================

*   Update tests for changes in retext-visit
*   Update jscs, retext, retext-content, retext-porter-stemmer, retext-visit
*   Update .gitignore, .npmignore, bower ignore
*   Move spec to test.js

0.1.1 / 2014-09-26
==================

*   Refactor Readme.md
*   Refactor API
*   Add npm deployment to travis
*   Remove `npm update` from travis
*   Remove browser test
*   Update .gitignore, .npmignore
*   Update Installation in docs
*   Add bower.json
*   Fix property order in component.json, package.json
*   Update copyright in Readme.md
*   Remove testling
*   Refactor spec
*   Update dependencies
*   Add components to gitignore

0.1.0 / 2014-07-13
==================

*   Added changelog
*   Fixed for API changes in retext
*   Update double-metaphone, eslint,istanbul, retext, retext-porter-stemmer, retext-visit
*   Removed functionality to browserify, complexity unit tests by default
*   Mention which metaphone implementation is used

0.0.3 / 2014-06-24
==================

*   Bring browserified spec up to date with [5a8fad7](https://github.com/wooorm/retext-double-metaphone/commit/5a8fad7)
*   Updated dependency version of jscs
*   Export the doubleMetaphone algorithm itself

0.0.2 / 2014-06-22
==================

*   Fixed Readme

0.0.1 / 2014-06-22
==================


test.js
'use strict';

/* eslint-env mocha */

/*
 * Dependencies.
 */

var assert = require('assert');
var retext = require('retext');
var nlcstToString = require('nlcst-to-string');
var visit = require('unist-util-visit');
var doubleMetaphone = require('./');

/*
 * Methods.
 */

var equal = assert.equal;
var dequal = assert.deepEqual;

/*
 * Fixtures.
 */

var sentence = 'A simple, English, sentence';

var phonetics = [
    ['A', 'A'],
    ['SMPL', 'SMPL'],
    ['ANKLX', 'ANLX'],
    ['SNTNS', 'SNTNS']
];

/**
 * Example stemmer, which expects the tree to equal `otherWord`.
 */
function stemmer() {
    return function (cst) {
        visit(cst, 'WordNode', function (node) {
            node.data = {
                'stem': nlcstToString(node)
            };
        });
    };
}

/*
 * Tests.
 */

describe('doubleMetaphone()', function () {
    var processor = retext.use(doubleMetaphone);

    it('should be of type `function`', function () {
        equal(typeof doubleMetaphone, 'function');
    });

    processor.process(sentence, function (e, file) {
        var cst = file.namespace('retext').cst;

        it('should not throw', function (done) {
            done(e);
        });

        it('should process each `WordNode`', function () {
            var index = -1;

            visit(cst, 'WordNode', function (node) {
                assert('phonetics' in node.data);
                dequal(node.data.phonetics, phonetics[++index]);
            });
        });
    });
});

describe('doubleMetaphone() with a stemmer', function () {
    var processor = retext.use(stemmer).use(doubleMetaphone);

    processor.process(sentence, function (e, file) {
        var cst = file.namespace('retext').cst;

        it('should not throw', function (done) {
            done(e);
        });

        it('should process `stem` in each `WordNode`', function () {
            visit(cst, 'WordNode', function (node) {
                dequal(node.data.stemmedPhonetics, node.data.phonetics);
            });
        });
    });
});


spec/retext-double-metaphone.spec.js
'use strict';

/**
 * Dependencies.
 */

var doubleMetaphone,
    content,
    stemmer,
    visit,
    Retext,
    assert;

doubleMetaphone = require('..');
Retext = require('retext');
visit = require('retext-visit');
content = require('retext-content');
stemmer = require('retext-porter-stemmer');
assert = require('assert');

/**
 * Retext.
 */

var retext,
    retextWithStemmer;

retext = new Retext()
    .use(doubleMetaphone)
    .use(visit)
    .use(content);

retextWithStemmer = new Retext()
    .use(doubleMetaphone)
    .use(stemmer)
    .use(visit)
    .use(content);

/**
 * Fixtures.
 */

var sentence,
    words,
    otherPhonetics,
    stemmedOtherPhonetics;

sentence = 'A simple, English, sentence.';

words = ['A', 'detestable', 'vile', 'paragraph'];

otherPhonetics = [
    ['A', 'A'],
    ['TTSTPL', 'TTSTPL'],
    ['FL', 'FL'],
    ['PRKRF', 'PRKRF']
];

stemmedOtherPhonetics = [
    ['A', 'A'],
    ['TTST', 'TTST'],
    ['FL', 'FL'],
    ['PRKRF', 'PRKRF']
];

/**
 * Tests.
 */

describe('double-metaphone()', function () {
    it('should be a `function`', function () {
        assert(typeof doubleMetaphone === 'function');
    });

    retext.parse(sentence, function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should process each `WordNode`', function () {
            tree.visitType(tree.WORD_NODE, function (node) {
                assert('phonetics' in node.data);
            });
        });

        it('should set `phonetics` to `null` when `WordNode` (no ' +
            'longer?) has a value',
            function () {
                tree.visitType(tree.WORD_NODE, function (node) {
                    node.removeContent();

                    assert(node.data.phonetics === null);
                });
            }
        );

        it('should re-process `WordNode`s when their value changes',
            function () {
                var index;

                index = -1;

                tree.visitType(tree.WORD_NODE, function (node) {
                    var phonetics;

                    index++;

                    node.replaceContent(words[index]);

                    phonetics = node.data.phonetics;

                    assert(phonetics[0] === otherPhonetics[index][0]);
                    assert(phonetics[1] === otherPhonetics[index][1]);
                });
            }
        );
    });
});

describe('double-metaphone() with a stemmer', function () {
    retextWithStemmer.parse(sentence, function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should process `stem` in each `WordNode`', function () {
            tree.visitType(tree.WORD_NODE, function (node) {
                assert('stemmedPhonetics' in node.data);
            });
        });

        it('should set `stemmedPhonetics` to `null` when `WordNode` (no ' +
            'longer?) has a value',
            function () {
                tree.visitType(tree.WORD_NODE, function (node) {
                    node.removeContent();

                    assert(node.data.stemmedPhonetics === null);
                });
            }
        );

        it('should re-process `WordNode`s when their stem changes',
            function () {
                var index;

                index = -1;

                tree.visitType(tree.WORD_NODE, function (node) {
                    var stemmedPhonetics;

                    index++;

                    node.replaceContent(words[index]);

                    stemmedPhonetics = node.data.stemmedPhonetics;

                    assert(
                        stemmedPhonetics[0] ===
                        stemmedOtherPhonetics[index][0]
                    );

                    assert(
                        stemmedPhonetics[1] ===
                        stemmedOtherPhonetics[index][1]
                    );
                });
            }
        );
    });
});


index.js
/**
 * @author Titus Wormer
 * @copyright 2014-2015 Titus Wormer
 * @license MIT
 * @module retext:double-metaphone
 * @fileoverview Retext implementation of the Double Metaphone algorithm.
 */

'use strict';

/*
 * Dependencies.
 */

var algorithm = require('double-metaphone');
var visit = require('unist-util-visit');
var nlcstToString = require('nlcst-to-string');

/**
 * Patch a `phonetics` property on `node` (a word-node).
 * When a `stem` property is available on the node’s data
 * object, a `stemmedPhonetics` is added on the data
 * as well.
 *
 * @param {NLCSTWordNode} node - Node.
 */
function patch(node) {
    var data = node.data || {};
    var value = nlcstToString(node);

    data.phonetics = algorithm(value);

    if ('stem' in data) {
        data.stemmedPhonetics = algorithm(data.stem);
    }

    node.data = data;
}

/**
 * Transformer.
 *
 * @param {NLCSTNode} cst - Syntax tree.
 */
function transformer(cst) {
    visit(cst, 'WordNode', patch);
}

/**
 * Attacher.
 *
 * @return {Function} - `transformer`.
 */
function attacher() {
    return transformer;
}

/*
 * Expose.
 */

module.exports = attacher;


benchmark.js
'use strict';

var Retext,
    doubleMetaphone;

/*
 * Dependencies.
 */

Retext = require('retext');
doubleMetaphone = require('./');

/*
 * Dependencies.
 */

var retext,
    retextWithDoubleMetaphone;

retext = new Retext();
retextWithDoubleMetaphone = new Retext().use(doubleMetaphone);

/*
 * Test data: A (big?) article (w/ 100 paragraphs, 500
 * sentences, 10,000 words);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

suite('retext w/o retext-double-metaphone', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retext.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retext.parse(section, done);
        }
    );
});

suite('retext w/ retext-double-metaphone', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retextWithDoubleMetaphone.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retextWithDoubleMetaphone.parse(section, done);
        }
    );
});
